lib/librt/sched.3
.\"	$NetBSD: sched.3,v 1.14.2.2 2016/11/04 14:48:54 pgoyette Exp $
.\"
.\" Copyright (c) 2008, 2016 The NetBSD Foundation, Inc.
.\" All rights reserved.
.\"
.\" This code is derived from software contributed to The NetBSD Foundation
.\" by Mindaugas Rasiukevicius <rmind at NetBSD org>.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
.\" ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
.\" TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
.\" PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
.\" BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\" SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
.\" INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
.\" CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
.\" ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd August 5, 2016
.Dt SCHED 3
.Os
.Sh NAME
.Nm sched_setparam ,
.Nm sched_getparam ,
.Nm sched_setscheduler ,
.Nm sched_getscheduler ,
.Nm sched_get_priority_max ,
.Nm sched_get_priority_min ,
.Nm sched_rr_get_interval ,
.Nm sched_yield ,
.Nm sched_protect ,
.Nm sched_setaffinity_np ,
.Nm sched_getaffinity_np
.Nd process scheduling
.Sh LIBRARY
.Lb librt
.Sh SYNOPSIS
.In sched.h
.Ft int
.Fn sched_setparam "pid_t pid" "const struct sched_param *param"
.Ft int
.Fn sched_getparam "pid_t pid" "struct sched_param *param"
.Ft int
.Fn sched_setscheduler "pid_t pid" "int policy" "const struct sched_param *param"
.Ft int
.Fn sched_getscheduler "pid_t pid"
.Ft int
.Fn sched_get_priority_max "int policy"
.Ft int
.Fn sched_get_priority_min "int policy"
.Ft int
.Fn sched_rr_get_interval "pid_t pid" "struct timespec *interval"
.Ft int
.Fn sched_yield "void"
.Ft int
.Fn sched_setaffinity_np "pid_t pid" "size_t size" "cpuset_t *cpuset"
.Ft int
.Fn sched_getaffinity_np "pid_t pid" "size_t size" "cpuset_t *cpuset"
.Ft int
.Fn sched_protect "int priority"
.Sh DESCRIPTION
This section describes the functions used to get scheduling information
about processes, and control the scheduling of processes.
.Pp
Available scheduling policies (classes) are:
.Bl -tag -width SCHED_OTHER
.It Dv SCHED_OTHER
Time-sharing (TS) scheduling policy.
The default policy in
.Nx .
.It Dv SCHED_FIFO
First in, first out (FIFO) scheduling policy.
.It Dv SCHED_RR
Round robin scheduling policy.
.El
.Pp
The
.Fa struct sched_param
contains at least one member:
.Bl -tag -width flags
.It Fa sched_priority
Specifies the priority of the process.
For
.Dv SCHED_OTHER ,
must be
.Dv PRI_NONE ;
the kernel will dynamically assign priorities to the thread based on
CPU load.
.El
.Sh FUNCTIONS
.Bl -tag -width compact
.It Fn sched_setparam pid param
Sets the scheduling parameters for the process specified by
.Fa pid
to
.Fa param .
If the value of
.Fa pid
is equal to zero, then the calling process is used.
.It Fn sched_getparam pid param
Gets the scheduling parameters of the process specified by
.Fa pid
into the structure
.Fa param .
If the value of
.Fa pid
is equal to zero, then the calling process is used.
.It Fn sched_setscheduler pid policy param
Set the scheduling policy and parameters for the process specified by
.Fa pid .
If the value of
.Fa pid
is equal to zero, then the calling process is used.
.It Fn sched_getscheduler pid
Returns the scheduling policy of the process specified by
.Fa pid .
If the value of
.Fa pid
is equal to zero, then the calling process is used.
.It Fn sched_get_priority_max policy
Returns the maximal priority which may be used for the scheduling policy
specified by
.Fa policy .
.It Fn sched_get_priority_min policy
Returns the minimal priority which may be used for the scheduling policy
specified by
.Fa policy .
.It Fn sched_rr_get_interval pid interval
Returns the time quantum into the structure
.Fa interval
of the process specified by
.Fa pid .
If the value of
.Fa pid
is equal to zero, then the calling process is used.
The process must be running at
.Fa SCHED_RR
scheduling policy.
.It Fn sched_yield
Yields a processor voluntarily and gives other threads a chance to run
without waiting for an involuntary preemptive switch.
.It Fn sched_setaffinity_np pid size cpuset
Set the affinity mask specified by
.Fa cpuset
for the process specified by
.Fa pid .
At least one valid CPU must be set in the mask.
.It Fn sched_getaffinity_np pid size cpuset
Get the affinity mask of the process specified by
.Fa pid
into the
.Fa cpuset .
.It Fn sched_protect priority
Performs priority protection using the
.Dv PTHREAD_PRIO_PROTECT
protocol.
This function will increase the protected priority of the caller thread to
.Fa priority
if the current thread's protected priority is smaller than
.Fa priority .
Multiple calls to
.Fn sched_protect
with a positive priority will
.Dq push
a priority level to the current thread, whereas calling
.Fn sched_protect
with a
.Fa priority
level of
.Dv \-1
will
.Dq pop
a priority level.
When the level reaches
.Dv 0
(the same number of
.Dq pushes
and
.Dq pops
have been issued) the original thread priority will be restored.
.El
.Sh IMPLEMENTATION NOTES
Setting CPU
.Xr affinity 3
requires super-user privileges.
Ordinary users can be allowed to control CPU affinity
of their threads via the
.Pa security.models.extensions.user_set_cpu_affinity
.Xr sysctl 7 .
See
.Xr secmodel_extensions 9 .
.Pp
Portable applications should not use the
.Fn sched_setaffinity_np
and
.Fn sched_getaffinity_np
functions.
.Sh RETURN VALUES
.Fn sched_protect ,
.Fn sched_setparam ,
.Fn sched_getparam ,
.Fn sched_rr_get_interval ,
and
.Fn sched_yield
return 0 on success.
Otherwise, \-1 is returned and
.Va errno
is set to indicate the error.
.Pp
.Fn sched_setscheduler
returns the previously used scheduling policy on success.
Otherwise, \-1 is returned and
.Va errno
is set to indicate the error.
.Pp
.Fn sched_getscheduler
returns the scheduling policy on success.
Otherwise, \-1 is returned and
.Va errno
is set to indicate the error.
.Pp
.Fn sched_get_priority_max
and
.Fn sched_get_priority_min
return the maximal/minimal priority value on success.
Otherwise, \-1 is returned and
.Va errno
is set to indicate the error.
.Pp
.Fn sched_setaffinity_np
and
.Fn sched_getaffinity_np
return 0 on success.
Otherwise, \-1 is returned and
.Va errno
is set to indicate the error.
.Sh ERRORS
The
.Fn sched_setparam
and
.Fn sched_setscheduler
functions fail if:
.Bl -tag -width Er
.It Bq Er EINVAL
At least one of the specified scheduling parameters was invalid.
.It Bq Er EPERM
The calling process has no appropriate privileges to perform the operation.
.It Bq Er ESRCH
No process can be found corresponding to the PID specified by
.Fa pid ,
and the value of
.Fa pid
is not zero.
.El
.Pp
The
.Fn sched_getparam
and
.Fn sched_getscheduler
functions fail if:
.Bl -tag -width Er
.It Bq Er EPERM
The calling process is not a super-user and its effective user id does not
match the effective user-id of the specified process.
.It Bq Er ESRCH
No process can be found corresponding to that specified by
.Fa pid ,
and the value of
.Fa pid
is not zero.
.El
.Pp
The
.Fn sched_get_priority_max
and
.Fn sched_get_priority_min
functions fail if:
.Bl -tag -width Er
.It Bq Er EINVAL
The specified scheduling policy is invalid.
.El
.Pp
The
.Fn sched_rr_get_interval
function fails if:
.Bl -tag -width Er
.It Bq Er ESRCH
No process can be found corresponding to that specified by
.Fa pid ,
and the value of
.Fa pid
is not zero.
.El
.Pp
The
.Fn sched_protect
function fails if:
.Bl -tag -width Er
.It Bq Er EINVAL
The thread was not priority protected.
.It Bq Er EPERM
The
.Fa priority
parameter was out of range (not in the range between
.Dv SCHED_PRIO_MIN
and
.Dv SCHED_PRIO_MAX ) .
.El
.Sh SEE ALSO
.Xr affinity 3 ,
.Xr cpuset 3 ,
.Xr pset 3 ,
.Xr schedctl 8
.Sh STANDARDS
These functions, except
.Fn sched_setaffinity_np
and
.Fn sched_getaffinity_np ,
are expected to conform the
.St -p1003.1-2001
standard.
.Sh HISTORY
The scheduling functions appeared in
.Nx 5.0 .


doc/CHANGES
# LIST OF CHANGES FROM LAST RELEASE:			<$Revision: 1.2208 $>
#
#
# [Note: This file does not mention every change made to the NetBSD source tree.
# For an exhaustive list, please subscribe to the `source-changes' mailing list,
# or see the mailing list archives.  For more information, send a message
# containing just the word `help' to `majordomo@NetBSD.org'.]
#
# ----------------------------------------------------------------------------
# PLEASE USE THE FOLLOWING STYLE WHEN ADDING ITEMS:
#
#	theme: Content ...
#		more content ...
#		more content. [developer DATE-FORMAT]
#
# That's it, one tab to start an item, and two tabs to indent the content.
# Also please start the content with a capital and end it with a dot.
#
# This is very important for the www scripts, otherwise it won't be able to
# parse the file and the HTML output will be mangled.
#
# See htutils/changes/changes2html script for more details.
#
# ----------------------------------------------------------------------------


Changes from NetBSD 7.0 to NetBSD 8.0:
	libc: Update to tzcode2014f. [christos 20140815]
	gmake:	Update to gmake-3.81 in external/gpl2/gmake and remove 
		gnu/dist/gmake (3.80) [christos 20140818]
	module(7): Create MODULAR_DEFAULT_AUTOLOAD config flag to control
		whether module autoloading is on by default.
		[jnemeth 20140824]
	kernel: Option SYMTAB_SPACE has been replaced with COPY_SYMTAB.
		The related changes to src/sys/conf require a rebuild of
		config(1). [joerg 20140824]
	zoneinfo: Import tzdata2014g. [apb 20140831]
	wm(4): Add internal SERDES support for 82575 and newer controllers.
		[msaitoh 20140903]
	or1k: Add new port.  [matt 20140903]
	awindma(4): Add support for Allwinner A10/A20 DMA controller.
		[jmcneill 20140906]
	awinac(4): Add support for Allwinner A10/A20 audio codec.
		[jmcneill 20140906]
	awinrtc(4): Add support for Allwinner A10/A20 real-time clock.
		[jmcneill 20140907]
	awinmmc(4): Use DMA for MMC transfers. [jmcneill 20140908]
	axp20x(4): Add AXP20x PMU driver. [jmcneill 20140909]
	rtsol(8): Removed in favour of dhcpcd. [roy 20140911]
	rtsold(8): Removed in favour of dhcpcd. [roy 20140911]
	dhcpcd(8): Import dhcpcd-6.4.5. [roy 20140918]
	arm: Add support for i.MX6 SoC. [ryo 20140925]
	gpt(8): Completed overhaul, including adding follwing subcommands:
		resize, set, unset, backup, restore, and resizedisk.
		[jnemeth 20140926]
	dhcpcd(8): Import dhcpcd-6.4.7. [roy 20140927]
	mpt(4): added bio(4) support. [jmcneill 20140927]
	gpt(8): added type subcommand to modify partition type.
		[jnemeth 20140927]
	zoneinfo: Import tzdata2014h. [apb 20141004]
	dhcpcd(8): Import dhcpcd-6.5.0. [roy 20141006]
	libc: Update to tzcode2014h. [christos 20141007]
	file(1): Upgraded to 5.20. [christos 20141010]
	arm: Add support for Allwinner A31 SoC. [jmcneill 20141010]
	awiniic(4): Add support for Allwinner A31 SoC. [jmcneill 20141012]
	awinp2wi(4): Add support for Allwinner A31 Push-Pull Two Wire
		Interface. [jmcneill 20141012]
	axp22x(4): Add AXP22x PMU driver. [jmcneill 20141012]
	awinrtc(4): Add support for Allwinner A31 real-time clock.
		[jmcneill 20141012]
	awindma(4): Add support for Allwinner A31 DMA controller.
		[jmcneill 20141013]
	awinac(4): Add support for Allwinner A31 audio codec.
		[jmcneill 20141013]
	motg(4): Add support for Allwinner A10/A20 SoC. [jmcneill 20141015]
	wpa: Import wpa_supplicant and hostapd 2.3. [christos 20141016]
	dhcpcd(8): Import dhcpcd-6.5.1. [roy 20141018]
	OpenSSH: Imported 6.7. [christos 20141018]
	resolvconf(8): Import openresolv-3.6.0 [roy 20141020]
	gcc: Support -fsanitize=address [christos 20141022]
	zoneinfo: Import tzdata2014i. [apb 20141023]
	libc: Update to tzcode2014i. [christos 20141023]
	openpam(3): update to 20140912 (ourouparia) [christos 20141024]
	pppd(8): updated to version 2.4.7. [christos 20141025]
	acpi(4): Updated ACPICA to 20140926. [christos 20141025]
	dhcpcd(8): Import dhcpcd-6.6.0 [roy 20141029]
	resolvconf(8): Import openresolv-3.6.1 [roy 20141029]
	rtsx(4): Add support for Realtek RTS5227/RTL8402/RTL8411/RTL8411B.
		[nonaka 20141029]
	flex(1): Import flex-2.5.37 [christos 20141029]
	iwn(4): Add support for Intel Centrino Wireless-N
		100/105/130/135/2200/2230. From OpenBSD. [nonaka 20141030]
	awge(4): Add support for gigabit ethernet found on various
		Allwinner SOCs. [martin 20141101]
	dhcpcd(8): Import dhcpcd-6.6.1. [roy 20141107]
	awinhdmi(4): Add support for Allwinner A20/A31 HDMI controller.
		[jmcneill 20141110]
	awintcon(4): Add support for Allwinner A20/A31 TV/LCD controller.
		[jmcneill 20141110]
	awindebe(4): Add support for Allwinner A20/A31 Display engine
		backend (DE-BE). [jmcneill 20141110]
	awinhdmiaudio(4): Add support for Allwinner A20/A31 HDMI audio
		controller. [jmcneill 20141111]
	dhcpcd(8): Import dhcpcd-6.6.2. [roy 20141114]
	awinir(4): Add support for Allwinner A20/A31 IR receiver.
		[jmcneill 20141115]
	zoneinfo: Import tzdata2014j. [apb 20141116]
	libc: Change arc4random(3) to use ChaCha20 and per-thread state.
		[riastradh 20141116]
	libc: Update to tzcode2014j. [christos 20141117]
	libpcap: Import 1.6.2. [christos 20141119]
	tcpdump(8): Import 4.6.2. [christos 20141119]
	dhcpcd(8): Import dhcpcd-6.6.4. [roy 20141126]
	dhcpcd(8): Import dhcpcd-6.6.5. [roy 20141209]
	bind: Import version 9.10.1-P1. [christos 20141209]
	dhcpcd(8): Import dhcpcd-6.6.6. [roy 20141217]
	powerpc: Switch to GCC 4.8. [matt 20141218]
	dhcpcd(8): Import dhcpcd-6.6.7. [roy 20141219]
	ntp: Import ntp 4.2.8. [christos 20141219]
	midirecord(1): Add program to record MIDI files.  [mrg 20141229]
	file: update to 5.22 security limits [christos 20150102]
	byacc: update to 20141128 for more bison compat [christos 20150103]
	mgx: driver for Southland Media Systems (now Quantum 3D) MGX SBus
		graphics devices [macallan 20150106]
	gcc: Import GCC 4.8.4.  [mrg 20150109]
	openssl: Import openssl 1.0.1k [spz 20150113]
	am-utils: Upgrade to 6.2; adds nfsv4, lustre (which we can't use)
		nfs3 rpc support for amfs [christos 20150117]

lib/libc/sys/sched.c
/*	$NetBSD: sched.c,v 1.2.10.1 2012/04/17 00:05:26 yamt Exp $	*/

/*
 * Copyright (c) 2008, Mindaugas Rasiukevicius <rmind at NetBSD org>
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
__RCSID("$NetBSD: sched.c,v 1.2.10.1 2012/04/17 00:05:26 yamt Exp $");

#include <string.h>
#include <unistd.h>
#include <errno.h>
#include <sched.h>
#include <signal.h>
#include <sys/param.h>
#include <sys/types.h>

/* All LWPs in the process */
#define	P_ALL_LWPS		0

/*
 * Scheduling parameters.
 */

int
sched_setparam(pid_t pid, const struct sched_param *param)
{
	struct sched_param sp;

	memset(&sp, 0, sizeof(struct sched_param));
	sp.sched_priority = param->sched_priority;
	return _sched_setparam(pid, P_ALL_LWPS, SCHED_NONE, &sp);
}

int
sched_getparam(pid_t pid, struct sched_param *param)
{

	return _sched_getparam(pid, P_ALL_LWPS, NULL, param);
}

int
sched_setscheduler(pid_t pid, int policy, const struct sched_param *param)
{
	struct sched_param sp;
	int ret, old_policy;

	ret = _sched_getparam(pid, P_ALL_LWPS, &old_policy, &sp);
	if (ret < 0)
		return ret;

	memset(&sp, 0, sizeof(struct sched_param));
	sp.sched_priority = param->sched_priority;
	ret = _sched_setparam(pid, P_ALL_LWPS, policy, &sp);
	if (ret < 0)
		return ret;

	return old_policy;
}

int
sched_getscheduler(pid_t pid)
{
	struct sched_param sp;
	int ret, policy;

	ret = _sched_getparam(pid, P_ALL_LWPS, &policy, &sp);
	if (ret < 0)
		return ret;

	return policy;
}

/*
 * Scheduling priorities.
 */

int
sched_get_priority_max(int policy)
{

	if (policy < SCHED_OTHER || policy > SCHED_RR) {
		errno = EINVAL;
		return -1;
	}
	return (int)sysconf(_SC_SCHED_PRI_MAX);
}

int
sched_get_priority_min(int policy)
{

	if (policy < SCHED_OTHER || policy > SCHED_RR) {
		errno = EINVAL;
		return -1;
	}
	return (int)sysconf(_SC_SCHED_PRI_MIN);
}

int
/*ARGSUSED*/
sched_rr_get_interval(pid_t pid, struct timespec *interval)
{

	if (pid && kill(pid, 0) == -1)
		return -1;
	interval->tv_sec = 0;
	interval->tv_nsec = sysconf(_SC_SCHED_RT_TS) * 1000;
	return 0;
}

/*
 * Process affinity.
 */

int
sched_getaffinity_np(pid_t pid, size_t size, cpuset_t *cpuset)
{

	return _sched_getaffinity(pid, P_ALL_LWPS, size, cpuset);
}

int
sched_setaffinity_np(pid_t pid, size_t size, cpuset_t *cpuset)
{

	return _sched_setaffinity(pid, P_ALL_LWPS, size, cpuset);
}
